DefaultWriteBehindProcessor improvement for partial failures at storeAll/deleteAll

dsukhoroslov opened the pull request:
- added agreed piece of code in the DefaultWriteBihindProcessor;
- added a test case in the WriteBehindFailAndRetryTest;

Fixes #9733

mdogan commented:
run-lab-run

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/6922/

mdogan commented:
@dsukhoroslov:
There's checkstyle failure;

>hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/DefaultWriteBehindProcessor.java:240:26: Anonymous inner class length is 33 lines (max allowed is 30).

You can test checkstyle on local. 
`mvn clean validate -P checkstyle`
See https://github.com/hazelcast/hazelcast#checkstyle-and-findbugs

dsukhoroslov commented:
@mdogan, I run it. The output is:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:2.15:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. 
Failed during checkstyle execution: There are 6038 errors reported by Checkstyle 6.1.1 with C:\Work\HZ\git\hazelcast/checkstyle/checkstyle.xml ruleset. 
```

mdogan commented:
@dsukhoroslov: interesting, checkstyle itself failed. 
Anyway, only failure is the one above. anonymous inner class length issue. You can either refactor anonymous class to a named inner class or somehow reduce its length.

dsukhoroslov commented:
@mdogan, I see two possible fixes for this checkstyle failure:
1) introduce new static class implementing RetryTask in the same file, move code at lines 240..272 in it;
2) just remove unnecessary comment on failureList method (remove lines 263..265);

which one do you prefer? (I prefer the second one as it keeps close RetryTask implementations for single and batch store operations)

mdogan commented:
@dsukhoroslov, fine with me. that comment looks redundant.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/DefaultWriteBehindProcessor.java:
```diff
@@ -243,14 +244,22 @@ private Map convertToObject(Map<Object, DelayedEntry> batchMap) {
+                    Iterator<Object> keys = batchMap.keySet().iterator();
+                    for (Object key = null; keys.hasNext(); key = keys.next()) {
+                        if (!map.containsKey(key)) {
+                            keys.remove();
```
Seems `keys.next()` is not called before `keys.remove()` for the first iteration.  In that case key is null.

dsukhoroslov replied:
hmm. Your observation looks correct. But then it should throw NPE at containsKey check, afaik. Ok, I'll trace it.

dsukhoroslov commented:
@ahmetmircik, fixed. The weird thing is that the test I added for this case has passed successfully from command line (mvn -Dtest=WriteBehindFailAndRetryTest test). Only when debugged test in IDE I found the batchMap contains binary data, and the test has failed properly..

ahmetmircik commented:
thanks it seems ok to me now, one last thing, can you please squash the commits into one and send a PR for master branch?

dsukhoroslov commented:
done, see #9747

ahmetmircik commented:
can you also please squash the commits of this PR as its master port? After that i will merge this PR.

ahmetmircik commented:
run-lab-run

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/6946/

ahmetmircik commented:
run-lab-run

devOpsHazelcast commented:
Test PASSed.

mdogan commented on hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/DefaultWriteBehindProcessor.java:
```diff
@@ -243,14 +244,22 @@ private Map convertToObject(Map<Object, DelayedEntry> batchMap) {
+                    result = operationType.processBatch(map, mapStore);
+                } catch (Exception ex) {
+                    Iterator<Object> keys = batchMap.keySet().iterator();
+                    while (keys.hasNext()) {
```
This change might be risky for a maintenance release if somehow a user clears the batch `map` (or removes entries unintentionally) that's passed to `MapStore.storeAll()/deleteAll()`. At the moment in javadoc, we don't mention anything about modification of batch. For instance, I cannot think a reason why someone implements it this way but for following `storeAll()` implementation;

```java
public void storeAll(Map<K, V> map) {
   try {
      [...]
   } catch (Exception e) {
      [...]       
   } finally {
      map.clear();
   }
}
```

@jerrinot: what do you think?

jerrinot replied:
I agree. It's a behavior change and we should avoid this in a maintenance release unless the old behavior is a clear bug. 
the [forward port is OK to me](https://github.com/hazelcast/hazelcast/pull/9747/files), assuming it also described the behavior in the contract of MapStore. 

@dsukhoroslov: the 3.8 is going to be out soon.

jerrinot commented:
@dsukhoroslov: I appreciate your effort, however it's unacceptable to change behavior like this in a maintanance release. 

Let's concentrate on the [PR into master](https://github.com/hazelcast/hazelcast/pull/9747/files).  That looks good to me, I believe this only missing bit is to describe the behaviour in the MapStore JavaDoc.
